ruby/sdk-lightweight.md

```diff
@@ -0,0 +1,314 @@
+# Ruby SDK (lightweight overview)
+
+This is a lightweight proposal for the Ruby SDK. The intention for this proposal is to give everyone
+an end-to-end feel for the Ruby SDK. Many aspects of this proposal will probably change in the more
+detailed versions later.
+
+
+## Types
+
+After many experiments we've decided to not proceed with Ruby typing at this point. This decision
+will be revisited when Ruby's typing system matures to support non-basic use-cases.
+
+
+## Defining an activity
+
+Activities are defined as classes ([a very common practice](https://guides.rubyonrails.org/active_job_basics.html)
+given that the language is object-oriented).
+
+`sum_activity.rb`
+
+```ruby
+class SumActivity < Temporal::Activity
+  # This will generate a default class-level stub SumActivity.execute to be used in workflows
+  def execute(a, b)
+    a + b
+  end
+end
+```
+
+#### Alternative approaches
+
+- Using plain methods as activities. This is not feasible in Ruby since non-namespaced methods
+  are global without a clear way to reference them (would have to be `method(:my_activity)`)
+
+- It is possible to allow multiple activities per class, for example:
+
+`calculator_activities.rb`
+
+```ruby
+class CalculatorActivities < Temporal::Activity
+  # Generate custom stubs on the class to be used in workflows:
+  # - CalculatorActivities.add
+  # - CalculatorActivities.subtract
+  activities :add, :subtract
+
+  def add(a, b)
+    a + b
+  end
+
+  def subtract(a, b)
+    a - b
+  end
+end
+```
+
+
+## Defining a workflow
+
+Similar to activities, workflows are also represented as classes.
+
+`sum_workflow.rb`
+
+```ruby
+require 'sum_activity'
+
+class SumWorkflow < Temporal::Workflow
+  def execute(a, b)
+    # - .execute! is a shorthand for .execute().await
+    # - only positional arguments are supported (inline with other languages), keyword
+    #   arguments are reserved for options
+    SumActivity.execute!(a, b, start_to_close_timeout: 10)
+  end
+end
+```
+
+#### Alternative approaches
+
+- For the same exact reasons it is not feasible to use non-namespaced methods for workflow
+
+- Similar to activities it should be possible to define multiple workflows per class, however this
+  will not be allowed to avoid a perception of a shared state between workflows
+
+
+## Start a workflow
+
+```ruby
+require 'temporal/client'
+require 'sum_workflow'
+
+connection = Temporal::Connection.new('localhost:7233')
+client = Temporal::Client.new(connection, namespace: 'demo')
+
+# workflow is referenced via a stub, not just a class
+handle = client.start_workflow(SumWorkflow.execute(1, 2), id: 'id-1', task_queue: 'demo')
```

Comment:
Can you confirm that if I called this activity method `add` or something, that this would be `SumWorkflow.add`? Meaning the activity invocation name will match the activity implementation name? (granted you may have to do a reverse map lookup to get the _actual_ string name if they overrode it)

Reply:
Yes, that's right, but this is a wrong line :)

Basically when invoking the workflow you would always have `SomeWorkflow.execute` method, but when calling an activity from a workflow the name would correspond to the name on your activity class:

```ruby
class MyActivities < Temporal::Activity
  activities :add, :subtract, 'custom-activity' => :divide
  ...
end

class MyWorkflow < Temporal::Workflow
  def execute
    MyActivities.add!(1, 2)

    # non-! version returns a future
    future = MyActivities.subtract(42, 10)
    
    # Schedules an activity with a type "custom-activity"
    MyActivities.divide!(4, 2)
  end
end
```